src/main/java/com/github/manolo8/darkbot/core/objects/swf/FlashMap.java

```diff
@@ -0,0 +1,379 @@
+
+        if (size <= 0 || size > MAX_SIZE || capacity <= 0 || capacity > MAX_CAPACITY) {
+            this.size = 0;
+            return;
+        }
+        if (entries.length < size) entries = Arrays.copyOf(entries, size);
+
+        API.readMemory(atoms, BUFFER, capacity); //TODO add readLongs in API
+        int currentSize = 0, realSize = 0;
+        for (int offset = 0; offset < capacity && currentSize < size; offset += 8) {
+            long keyAtom = ByteUtils.getLong(BUFFER, offset);
+
+            if (keyAtom == EMPTY) continue;
+            if (keyAtom == DELETED) {
+                offset += 8;
+                currentSize++;
+                continue; // skip deleted pair
+            }
+
+            AtomKind keyKind = AtomKind.of(keyAtom);
+            if (keyKind != this.keyKind) {
+                System.out.println("Invalid keyKind! expected: " + this.keyKind + ", read: " + keyKind);
+                break;
+            }
+
+            long valueAtom = ByteUtils.getLong(BUFFER, (offset += 8));
+            AtomKind valueKind = AtomKind.of(valueAtom);
+            if (valueKind != this.valueKind) {
+                System.out.println("Invalid valueKind! expected: " + this.valueKind + ", read: " + valueKind);
+                break;
+            }
+
+            Entry entry = entries[realSize];
+            if (entry == null)
+                entry = entries[realSize] = new Entry();
+
+            entry.set(keyAtom, valueAtom);
+
+            realSize++;
+            currentSize++;
+        }
+
+        for (int i = realSize; i < this.size; i++) {
+            Entry entry = entries[i];
+            if (entry != null) entry.reset();
+        }
+
+        if (updatables != null)
+            updatables.forEach((k, wrapper) -> wrapper.resetIfNoReferences());
+
+        this.size = realSize;
+    }
+
+    @Override
+    public void clear() {
+        this.size = 0;
+        //noinspection unchecked
+        this.entries = (Entry[]) Array.newInstance(Entry.class, 0);
+        if (updatables != null)
+            updatables.forEach((k, wrapper) -> wrapper.resetReferences());
+    }
+
+    @NotNull
+    @Override
+    public Set<Map.Entry<K, V>> entrySet() {
+        return new EntrySet();
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public V get(Object key) {
+        int i = indexOf(key);
+        return i == -1 ? null : entries[i].value;
+    }
+
+    @Override
+    public V put(K key, V value) {
+        if (!valueUpdatable) throw new IllegalStateException("V Type must be Updatable!");
+        if (updatables == null) updatables = new HashMap<>();
+        updatables.put(key, new ValueWrapper(value));
+
+        return value;
+    }
+
+    @Override
+    public V remove(Object key) {
+        if (updatables == null) return null;
+        ValueWrapper v = updatables.remove(key);
+        return v == null ? null : v.value;
```

Comment:
This may violate some contract, you can call `get(a)` and get a response, but call `remove(a)` and get a null returned, if this is not a map of updatables